pkg/csi/service/node.go

```diff
@@ -109,8 +111,32 @@ func (driver *vsphereCSIDriver) NodeUnstageVolume(
 	log := logger.GetLogger(ctx)
 	log.Infof("NodeUnstageVolume: called with args %+v", *req)
 
+	// Validate arguments
+	volumeID := req.GetVolumeId()
 	stagingTarget := req.GetStagingTargetPath()
 
+	if len(volumeID) == 0 {
+		return nil, logger.LogNewErrorCode(log, codes.InvalidArgument, "NodeUnstageVolume Volume ID must be provided")
+	}
+	if len(stagingTarget) == 0 {
+		return nil, logger.LogNewErrorCode(log, codes.InvalidArgument,
+			"NodeUnstageVolume Staging Target Path must be provided")
+	}
+
+	// There have few observations where the 1st NodeUnstageVolume call takes more time and remains in-progress for
+	// longer durations. However, k8s will issue a 2nd NodeUnstageVolume call assuming the 1st NodeUnstageVolume
+	// has timed out. In such cases, the 2nd NodeUnstageVolume call succeeds as the target Mountpoint is not found.
+	// Therefore, a DetachVolume will be invoked while the 1st NodeUnstageVolume is still in-progress and in-turn
+	// corrupts the volume. To avoid the such volume corruption issue, we can keep a lock per VolumeID during the
+	// NodeUnstageVolume operation.
+	actual, _ := driver.volumeLocks.LoadOrStore(volumeID, &sync.Mutex{})
```

Comment:
This will grow infinitely. You would have to evict keys somehow, you were sort of doing it in your previous changes.

Reply:
Added `driver.volumeLocks.Delete(volumeID)` statement, which will ensure we delete the key after the 1st call is processed. And also, the delete() internally ensures the key is removed if it is present and also unlocks the mutex regardless of whether the key was present. This behavior is from the Sync/map package.

Reply:
if we delete the lock from store, and another thread is already waiting to get it unlocked. Do you see any issue in another thread? Can we confirm node driver is not crashing due to this?

Reply:
The Delete() method in Sync.map manages all of the necessary locking and synchronization internally, ensuring that operations are thread-safe and free from concurrency issues. I don't think there is any issue with the second thread when first thread deletes lock

Reply:
Assume`driver.volumeLocks.Delete(volumeID)` is executed by first thread and another thread has found the mutex in the map and already waiting at `volumeLock.Lock()`
And while second thread gets lock another request comes and it does not find volumeid in the map and creates another lock.